docs/reference/content/upgrading.md

```diff
@@ -7,11 +7,39 @@ title = "Upgrade Considerations"
   pre = "<i class='fa fa-level-up'></i>"
 +++
 
-## Upgrading from 3.x
-
-... TODO ...
-
-## Upgrading from mongo-java-driver-reactivestreams 1.12
+# Upgrading to the 4.0 Driver
+
+## Upgrading from the 3.12 Java driver
+
+The 4.0 release is a major release as per the definition of semantic versioning included in [semver.org]. As such, users
+that upgrade to this release should expect breaking changes. That said, we have attempted to ensure that the upgrade 
+process is as seamless as possible.  Breaking changes are as follows:
+
+  * Numerous classes and methods have been removed from the driver. All of these API elements are annotated as deprecated in the 
+    3.12 release, so if you compile your application with 3.12 and enable deprecation warnings in the compiler, you will be
+    able to locate all uses of these API elements and follow the recommendations contained in the Javadoc for each API element
+    to remove usage from your application. Most of these API elements are unlikely to be used by normal applications, but one bears
+    mentioning explicitly: the entire callback-baked asynchronous driver has been removed. Applications relying on this driver must either
+    port their application to the Reactive Streams driver, or else must remain on a 3.x driver release.
+  * The insert helper methods now return an insert result object instead of void. While this does not break source compatibility, it does
+    break binary compatibility, so any class that calls one of the insert helper methods must be recompiled with the 4.0 driver.
```

Comment:
I'm not sure we need to mention binary/source compatibility here.  

Given the extent of the changes we should clarify that users are expected to recompile against the 4.0 version of the driver.

Reply:
I do believe that this is the only change that requires recompilation.  But as nearly every application calls insert, it's sort of irrelevant.  I'll add a separate note saying that recompilation is required.

Reply:
Done.